### DISCUSSION TITLE
Extending the MS Hex Editor

### DISCUSSION THREAD

**Discussion by jw3** [jw3/example-daffodil-vscode#145] [Design Discussions]
Microsoft has a Hex Editor extension that looks like a nice potential starting point

- https://marketplace.visualstudio.com/items?itemName=ms-vscode.hexeditor&ssr=false
- https://github.com/microsoft/vscode-hexeditor

Questions
1. How is the performance on large files?
2. Can we extend this extension by consuming it as an artifact?
3. Does this support binary and text files?
4. Can we extend this by adding menu options?

**Comment by shanedell** [comment]
1. Performance on large files seems to work well. I tested with a 20MB image and loaded within seconds.
2. We can extend the extension by adding it as a `extensionDependencies` inside of the `package.json`

Testing with extending:
- There is a way to make it open a desired file to the side.
- It seems that you only get the commands from the extension you don't get its types.
- Since it uses web-views to display to file there are some issues:
  - We cannot in our extensions code access/grab the web-views the `vscode-hexeditor` creates.
  - The previous issue causes another issue of not being able to highlight the current byte that the debugger is at.
 
Possible Fixes to above issues:
- Fork the extension make, make changes we want and publish the extension into the marketplace for this the debugger extension to be able to use it.
  - This may not fix the issues with not being able to grab the web-views
- Come up with a script, would have to be JavaScript, that could be fired whenever the hex-view is opened that tries to grab the elements from the web-view document.
  - Not sure if this would work just an idea

**Comment by scholarsmate** [comment]
I tried it against a nearly 1 GiB zip file.  It complained that loading such a large file might lead to instability, but I loaded it anyway.  It took several seconds to read in the file and render to the main editor window.  I scrolled around randomly and dragged the scrollbar to the bottom to make sure I can view the entire file.  I then edited (overwrite) a few bytes in the middle of the file and the extension host crashed.  Here is the traceback I received:

```
<--- Last few GCs --->
[58105:0x1e00000000] 149481917 ms: Scavenge 1080.0 (1092.0) -> 1072.3 (1092.0) MB, 0.2 / 0.0 ms  (average mu = 0.998, current mu = 0.929) allocation failure 
[58105:0x1e00000000] 149481927 ms: Scavenge 1080.0 (1092.0) -> 1072.3 (1092.0) MB, 0.2 / 0.0 ms  (average mu = 0.998, current mu = 0.929) allocation failure 
[58105:0x1e00000000] 149481938 ms: Scavenge 1080.0 (1092.0) -> 1072.3 (1092.0) MB, 0.2 / 0.0 ms  (average mu = 0.998, current mu = 0.929) allocation failure 
<--- JS stacktrace --->
FATAL ERROR: invalid array length Allocation failed - JavaScript heap out of memory
 1: 0x111958a35 node::Buffer::New(v8::Isolate*, char*, unsigned long, void (*)(char*, void*), void*) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
 2: 0x111958bd6 node::OnFatalError(char const*, char const*) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
 3: 0x11262a023 v8::internal::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
 4: 0x112629f7f v8::internal::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
 5: 0x1126d56c5 v8::internal::Heap::PreciseCollectAllGarbage(int, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
 6: 0x10f335216 v8::internal::Factory::NewUninitializedFixedArray(int) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
 7: 0x10f4ad9e5 v8::internal::ElementsKindToByteSize(v8::internal::ElementsKind) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
 8: 0x10f4ad246 v8::internal::ElementsKindToByteSize(v8::internal::ElementsKind) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
 9: 0x10f5084d2 v8::internal::JSObject::AddDataElement(v8::internal::Handle<v8::internal::JSObject>, unsigned int, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyAttributes) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
10: 0x10f55be80 v8::internal::Object::SetProperty(v8::internal::LookupIterator*, v8::internal::Handle<v8::internal::Object>, v8::internal::StoreOrigin, v8::Maybe<v8::internal::ShouldThrow>) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
11: 0x10f4f7ff7 v8::internal::JSReceiver::GetOrCreateIdentityHash(v8::internal::Isolate*) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
12: 0x10f659bca v8::internal::Runtime::SetObjectProperty(v8::internal::Isolate*, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, v8::internal::StoreOrigin, v8::Maybe<v8::internal::ShouldThrow>) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
13: 0x1e000b22b8 
14: 0x1e000dbc7b 
15: 0x1e000c5335 
16: 0x1e00049541 
```

I also can't find a way to insert bytes using this extension.
